analyzer/src/main/kotlin/managers/Carthage.kt

```diff
@@ -0,0 +1,282 @@
+    class Factory : AbstractPackageManagerFactory<Carthage>("Carthage") {
+        // TODO: Add support for the Cartfile.
+        //  This would require to resolve the actual dependency versions as a Cartfile supports dynamic versions.
+        override val globsForDefinitionFiles = listOf("Cartfile*.resolved")
+
+        override fun create(
+            analysisRoot: File,
+            analyzerConfig: AnalyzerConfiguration,
+            repoConfig: RepositoryConfiguration
+        ) = Carthage(managerName, analysisRoot, analyzerConfig, repoConfig)
+    }
+
+    override fun resolveDependencies(definitionFile: File): List<ProjectAnalyzerResult> {
+        // Transitive dependencies are only supported if the the dependency itself uses Carthage.
+        // See: https://github.com/Carthage/Carthage#nested-dependencies
+        val workingDir = definitionFile.parentFile
+        val (user, project, ref) = getProjectInfoFromVcs(workingDir)
+        return listOf(
+            ProjectAnalyzerResult(
+                project = Project(
+                    id = Identifier(
+                        type = managerName,
+                        namespace = user.orEmpty(),
+                        name = project.orEmpty(),
+                        version = ref.orEmpty()
+                    ),
+                    definitionFilePath = VersionControlSystem.getPathInfo(definitionFile).path,
+                    declaredLicenses = sortedSetOf(),
+                    vcs = VcsInfo.EMPTY,
+                    vcsProcessed = processProjectVcs(workingDir, VcsInfo.EMPTY),
+                    scopes = sortedSetOf(),
+                    homepageUrl = ""
+                ),
+                packages = parseCarthageDependencies(definitionFile)
+            )
+        )
+    }
+
+    /**
+     * As the Carthage.resolved file does not provide any project information, trying to retrieve some from VCS.
+     */
+    private fun getProjectInfoFromVcs(workingDir: File): Triple<String?, String?, String?> {
+        val workingTree = VersionControlSystem.forDirectory(workingDir)
+        val vcsInfo = workingTree?.getInfo() ?: VcsInfo.EMPTY
+        val normalizedVcsUrl = normalizeVcsUrl(vcsInfo.url)
+        val vcsHost = VcsHost.fromUri(URI(normalizedVcsUrl))
+        return Triple(
+            vcsHost?.getUserOrOrganization(normalizedVcsUrl),
+            vcsHost?.getProject(normalizedVcsUrl),
+            vcsInfo.resolvedRevision
+        )
+    }
+
+    private fun parseCarthageDependencies(definitionFile: File): SortedSet<Package> {
+        val dependencyLines = definitionFile.readLines()
+        val workingDir = definitionFile.parent
+        val packages = sortedSetOf<Package>()
+        dependencyLines.forEach { line ->
+            if (line.isNotEmpty() && !isComment(line)) {
+                val (carthageType, id, ref) = parseCarthageDependencyLine(line)
+                when (carthageType) {
```

Comment:
Would it make sense to move this whole `when` block into `parseCarthageDependencyLine()` and make that function return a package?

...Amongst other this would remove the redundant `packages.add()` calls.

Reply:
Done. This makes the calling code pretty clean: `packages.add(parseDependencyLine(line, workingDir))`
I'm not a fan of passing `workingDir` to that method, but not sure yet how to get rid of it.

Reply:
Speaking of `packages.add()`, please use the `+=` operator instead where possible (also see my recent #3093).